File: practicalex-2/src/main/java/edu/multimedia/lucene/search/MovieSearcher.java
package edu.multimedia.lucene.search;

import edu.multimedia.lucene.index.MovieIndexer;
import org.apache.lucene.analysis.standard.StandardAnalyzer;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.DoublePoint;
import org.apache.lucene.document.IntPoint;
import org.apache.lucene.index.DirectoryReader;
import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
import org.apache.lucene.queryparser.classic.ParseException;
import org.apache.lucene.search.*;
import org.apache.lucene.store.Directory;
import org.apache.lucene.store.FSDirectory;

import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Provides search functionality over the Lucene index.
 * Supports basic keyword search and enhanced search with filters.
 */
public class MovieSearcher {
    private final Path indexDirectory;
    private final StandardAnalyzer analyzer;
    private DirectoryReader reader;
    private IndexSearcher searcher;

    public MovieSearcher(Path indexDirectory) throws IOException {
        this.indexDirectory = indexDirectory;
        this.analyzer = new StandardAnalyzer();
        openIndex();
    }

    /**
     * Opens the index for searching.
     */
    private void openIndex() throws IOException {
        Directory directory = FSDirectory.open(indexDirectory);
        reader = DirectoryReader.open(directory);
        searcher = new IndexSearcher(reader);
    }

    /**
     * Performs a basic keyword search across title, overview, and cast fields.
     * 
     * @param queryText Search query
     * @param maxResults Maximum number of results to return
     * @return List of search results
     * @throws ParseException If query parsing fails
     * @throws IOException If search fails
     */
    public List<SearchResult> basicSearch(String queryText, int maxResults) 
            throws ParseException, IOException {
        // Check searcher state
        if (searcher == null || reader == null) {
            openIndex();
        }
        
        // Search in multiple fields with different boosts
        String[] fields = {
            MovieIndexer.FIELD_TITLE,
            MovieIndexer.FIELD_CAST,
            MovieIndexer.FIELD_OVERVIEW,
            MovieIndexer.FIELD_TAGLINE,
            MovieIndexer.FIELD_GENRES
        };
        
        Map<String, Float> boosts = new HashMap<>();
        boosts.put(MovieIndexer.FIELD_TITLE, 2.0f);      // Title matches are most important
        boosts.put(MovieIndexer.FIELD_CAST, 1.5f);       // Cast matches are important
        boosts.put(MovieIndexer.FIELD_OVERVIEW, 1.0f);
        boosts.put(MovieIndexer.FIELD_TAGLINE, 1.0f);
        boosts.put(MovieIndexer.FIELD_GENRES, 1.0f);
        
        MultiFieldQueryParser parser = new MultiFieldQueryParser(fields, analyzer, boosts);
        Query query = parser.parse(queryText);
        
        return executeSearch(query, null, maxResults);
    }

    /**
     * Performs an enhanced search with filters for rating, year, etc.
     * 
     * @param queryText Search query
     * @param minRating Minimum rating (null = no filter)
     * @param maxRating Maximum rating (null = no filter)
     * @param minYear Minimum year (null = no filter)
     * @param maxYear Maximum year (null = no filter)
     * @param maxResults Maximum number of results to return
     * @return List of search results
     * @throws ParseException If query parsing fails
     * @throws IOException If search fails
     */
    public List<SearchResult> enhancedSearch(String queryText, 
                                           Double minRating, Double maxRating,
                                           Integer minYear, Integer maxYear,
                                           int maxResults) 
            throws ParseException, IOException {
        // Check searcher state
        if (searcher == null || reader == null) {
            openIndex();
        }
        
        // Build text query
        String[] fields = {
            MovieIndexer.FIELD_TITLE,
            MovieIndexer.FIELD_CAST,
            MovieIndexer.FIELD_OVERVIEW,
            MovieIndexer.FIELD_TAGLINE,
            MovieIndexer.FIELD_GENRES
        };
        
        Map<String, Float> boosts = new HashMap<>();
        boosts.put(MovieIndexer.FIELD_TITLE, 2.0f);
        boosts.put(MovieIndexer.FIELD_CAST, 1.5f);
        boosts.put(MovieIndexer.FIELD_OVERVIEW, 1.0f);
        boosts.put(MovieIndexer.FIELD_TAGLINE, 1.0f);
        boosts.put(MovieIndexer.FIELD_GENRES, 1.0f);
        
        MultiFieldQueryParser parser = new MultiFieldQueryParser(fields, analyzer, boosts);
        Query textQuery = parser.parse(queryText);
        
        // Build filter query
        List<Query> filterQueries = new ArrayList<>();
        
        if (minRating != null || maxRating != null) {
            double min = minRating != null ? minRating : 0.0;
            double max = maxRating != null ? maxRating : 10.0;
            Query ratingQuery = DoublePoint.newRangeQuery(
                MovieIndexer.FIELD_RATING, min, max);
            filterQueries.add(ratingQuery);
        }
        
        if (minYear != null || maxYear != null) {
            int min = minYear != null ? minYear : 1900;
            int max = maxYear != null ? maxYear : 2100;
            Query yearQuery = IntPoint.newRangeQuery(
                MovieIndexer.FIELD_YEAR, min, max);
            filterQueries.add(yearQuery);
        }
        
        Query filterQuery = null;
        if (!filterQueries.isEmpty()) {
            if (filterQueries.size() == 1) {
                filterQuery = filterQueries.get(0);
            } else {
                // Combine filters with AND
                filterQuery = new BooleanQuery.Builder()
                    .add(filterQueries.get(0), BooleanClause.Occur.MUST)
                    .add(filterQueries.get(1), BooleanClause.Occur.MUST)
                    .build();
            }
        }
        
        return executeSearch(textQuery, filterQuery, maxResults);
    }

    /**
     * Executes a search query with optional filter.
     * 
     * @param query Main search query
     * @param filter Optional filter query
     * @param maxResults Maximum number of results
     * @return List of search results
     * @throws IOException If search fails
     */
    private List<SearchResult> executeSearch(Query query, Query filter, int maxResults) 
            throws IOException {
        List<SearchResult> results = new ArrayList<>();
        
        // Check if reader is open
        if (reader == null || searcher == null) {
            System.err.println("Warning: Index reader not initialized, reopening...");
            openIndex();
        }
        
        // Apply filter if provided
        Query finalQuery = query;
        if (filter != null) {
            finalQuery = new BooleanQuery.Builder()
                .add(query, BooleanClause.Occur.MUST)
                .add(filter, BooleanClause.Occur.FILTER)
                .build();
        }
        
        // Execute search
        TopDocs topDocs = searcher.search(finalQuery, maxResults);
        
        // Extract results
        for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
            Document doc = searcher.doc(scoreDoc.doc);
            SearchResult result = new SearchResult(doc, scoreDoc.score);
            results.add(result);
        }
        
        return results;
    }

    /**
     * Refreshes the index reader (call after re-indexing).
     */
    public void refresh() throws IOException {
        if (reader != null) {
            reader.close();
        }
        // Reopen the index
        openIndex();
    }

    /**
     * Closes the searcher and releases resources.
     */
    public void close() throws IOException {
        if (reader != null) {
            reader.close();
        }
        analyzer.close();
    }
}

File: practicalex-2/whoosh-implementation/movie_searcher.py
"""
Whoosh-based movie searcher.
Implements basic and enhanced search functionality.
"""
from whoosh import index
from whoosh.qparser import MultifieldParser, AndGroup
from whoosh.query import And, Or, Term, NumericRange
from pathlib import Path


class MovieSearcher:
    """Provides search functionality over the Whoosh index."""
    
    def __init__(self, index_dir):
        """Initialize searcher with index directory."""
        self.index_dir = Path(index_dir)
        if not index.exists_in(str(self.index_dir)):
            raise ValueError(f"Index not found at {self.index_dir}")
        self.ix = index.open_dir(str(self.index_dir))
    
    def basic_search(self, query_text, max_results=10):
        """
        Performs basic keyword search across multiple fields.
        
        Args:
            query_text: Search query string
            max_results: Maximum number of results to return
            
        Returns:
            List of search results (dictionaries)
        """
        with self.ix.searcher() as searcher:
            # Define fields to search with boosts
            fields = ['title', 'cast', 'overview', 'tagline', 'genres']
            
            # Create multifield parser with boosts (fieldboosts is a dict parameter)
            fieldboosts = {
                'title': 2.0,   # Title matches are most important
                'cast': 1.5     # Cast matches are important
            }
            parser = MultifieldParser(fields, schema=self.ix.schema, fieldboosts=fieldboosts)
            
            # Parse query
            query = parser.parse(query_text)
            
            # Execute search
            results = searcher.search(query, limit=max_results)
            
            # Extract results
            search_results = []
            for hit in results:
                result = {
                    'imdb_id': hit['imdb_id'],
                    'title': hit['title'],
                    'year': hit.get('year', 0),
                    'rating': hit.get('rating', 0.0),
                    'overview': hit.get('overview', ''),
                    'cast': hit.get('cast', ''),
                    'score': hit.score
                }
                search_results.append(result)
            
            return search_results, results.scored_length()
    
    def enhanced_search(self, query_text, min_rating=None, max_rating=None,
                       min_year=None, max_year=None, max_results=10):
        """
        Performs enhanced search with filters.
        
        Args:
            query_text: Search query string
            min_rating: Minimum rating (optional)
            max_rating: Maximum rating (optional)
            min_year: Minimum year (optional)
            max_year: Maximum year (optional)
            max_results: Maximum number of results
            
        Returns:
            List of search results (dictionaries)
        """
        with self.ix.searcher() as searcher:
            # Build text query
            fields = ['title', 'cast', 'overview', 'tagline', 'genres']
            fieldboosts = {
                'title': 2.0,   # Title matches are most important
                'cast': 1.5      # Cast matches are important
            }
            parser = MultifieldParser(fields, schema=self.ix.schema, fieldboosts=fieldboosts)
            text_query = parser.parse(query_text)
            
            # Build filter queries
            filter_queries = []
            
            if min_rating is not None or max_rating is not None:
                min_r = min_rating if min_rating is not None else 0.0
                max_r = max_rating if max_rating is not None else 10.0
                rating_query = NumericRange('rating', min_r, max_r)
                filter_queries.append(rating_query)
            
            if min_year is not None or max_year is not None:
                min_y = min_year if min_year is not None else 1900
                max_y = max_year if max_year is not None else 2100
                year_query = NumericRange('year', min_y, max_y)
                filter_queries.append(year_query)
            
            # Combine queries
            if filter_queries:
                final_query = And([text_query] + filter_queries)
            else:
                final_query = text_query
            
            # Execute search
            results = searcher.search(final_query, limit=max_results)
            
            # Extract results
            search_results = []
            for hit in results:
                result = {
                    'imdb_id': hit['imdb_id'],
                    'title': hit['title'],
                    'year': hit.get('year', 0),
                    'rating': hit.get('rating', 0.0),
                    'overview': hit.get('overview', ''),
                    'cast': hit.get('cast', ''),
                    'score': hit.score
                }
                search_results.append(result)
            
            return search_results, results.scored_length()


def display_results(results, total_hits):
    """Display search results in a formatted way."""
    print(f"\n--- Search Results ({len(results)} of {total_hits} found) ---")
    
    if not results:
        print("No results found.")
        return
    
    for i, result in enumerate(results, 1):
        overview = result['overview']
        if len(overview) > 100:
            overview = overview[:100] + "..."
        
        print(f"\n{i}. Score: {result['score']:.3f} | {result['title']} ({result['year']}) | "
              f"Rating: {result['rating']:.1f} | {overview}")


if __name__ == "__main__":
    # Example usage
    index_path = "whoosh_index"
    
    searcher = MovieSearcher(index_path)
    
    # Basic search
    print("=== Basic Search ===")
    results, total = searcher.basic_search("star wars", max_results=10)
    display_results(results, total)
    
    # Enhanced search
    print("\n=== Enhanced Search ===")
    results, total = searcher.enhanced_search(
        "action", 
        min_rating=7.0,
        min_year=1990,
        max_results=10
    )
    display_results(results, total)

